rhino/src/main/java/org/mozilla/javascript/regexp/NativeRegExpStringIterator.java

```diff
@@ -0,0 +1,111 @@
+    private static final long serialVersionUID = 1L;
+    private static final String ITERATOR_TAG = "RegExpStringIterator";
+
+    public static void init(ScriptableObject scope, boolean sealed) {
+        ES6Iterator.init(scope, sealed, new NativeRegExpStringIterator(), ITERATOR_TAG);
+    }
+
+    /** Only for constructing the prototype object. */
+    private NativeRegExpStringIterator() {
+        super();
+    }
+
+    public NativeRegExpStringIterator(
+            Scriptable scope,
+            Scriptable regexp,
+            String string,
+            boolean global,
+            boolean fullUnicode) {
+        super(scope, ITERATOR_TAG);
+
+        this.regexp = regexp;
+        this.string = string;
+        this.global = global;
+        this.fullUnicode = fullUnicode;
+        this.nextDone = false;
+    }
+
+    @Override
+    public String getClassName() {
+        return "RegExp String Iterator";
+    }
+
+    @Override
+    protected boolean isDone(Context cx, Scriptable scope) {
+        // The base class calls _first_ isDone and _then_ nextValue, so we'll just compute the next
+        // value here and return it form "nextValue".
+        // Also, for non-global regexp, we need to return the first match and then "done" on the
+        // next iteration.
+
+        if (nextDone) {
+            return true;
+        }
+
+        next = regExpExec(cx, scope);
+        if (next == null) {
+            // Done! Point ii of the spec
+            next = Undefined.instance;
+            nextDone = true;
+            return true;
+        } else if (!global) {
+            // Return false at this iteration, but true at the next. Point iii of the spec
+            nextDone = true;
+            return false;
+        }
+
+        // Increment index if matched empty string, as per the spec, point v.
+        String matchStr = ScriptRuntime.toString(ScriptRuntime.getObjectIndex(next, 0, cx, scope));
+        if (matchStr.isEmpty()) {
+            long thisIndex =
+                    ScriptRuntime.toLength(ScriptRuntime.getObjectProp(regexp, "lastIndex", cx));
+            long nextIndex = ScriptRuntime.advanceStringIndex(string, thisIndex, fullUnicode);
+            ScriptRuntime.setObjectProp(regexp, "lastIndex", nextIndex, cx);
+        }
+
+        return false;
+    }
+
+    @Override
+    protected Object nextValue(Context cx, Scriptable scope) {
+        return next;
+    }
+
+    private Object regExpExec(Context cx, Scriptable scope) {
+        // See ECMAScript spec 22.2.7.1
+        Object execMethod = ScriptRuntime.getObjectProp(regexp, "exec", cx);
+        if (execMethod instanceof Callable) {
+            return ((Callable) execMethod).call(cx, scope, regexp, new Object[] {string});
+        }
+        return NativeRegExp.js_exec(cx, scope, regexp, new Object[] {string});
+    }
+
+    @Override
+    protected String getTag() {
+        return ITERATOR_TAG;
+    }
+
+    private Scriptable regexp;
```

Comment:
FWIW, the Rhino creators in the Old Days seemed to like to put variables at the end of the class, but for new stuff I think it's OK (and preferable) if we act like modern-day Java developers and put them at the beginning.

Reply:
Done